### PR TITLE
Evaluator

### PR DESCRIPTION
**Pull request by ibralibya**

### CODE REVIEW

File: lib/collaborative_filtering.py [hunk 1]
```diff
@@ -52,24 +52,25 @@ def set_config(self, config):
...
         train = self.ratings.copy()
+
         # TODO split in a more intelligent way
         for user in range(self.ratings.shape[0]):
             non_zeros = self.ratings[user, :].nonzero()[0]
             test_ratings = numpy.random.choice(non_zeros,
                                                size=int(test_percentage * len(non_zeros)))
             train[user, test_ratings] = 0.
             test[user, test_ratings] = self.ratings[user, test_ratings]
-        assert(numpy.all((train * test) == 0))
```
Comment (ndomar) [inline lib/collaborative_filtering.py]: Why remove this line? It is normal to check here if the data is not the same.

File: tests/collaborative_tests.py [hunk 2]
```diff
@@ -51,6 +51,11 @@ def runTest(self):
         random_item = int(numpy.random.random() * self.documents)
         random_prediction = cf.predict(random_user, random_item)
         self.assertTrue(isinstance(random_prediction, numpy.float64))
+        train, test = cf.naive_split(test_percentage=0.2)
+        self.assertEqual(numpy.count_nonzero(train) + numpy.count_nonzero(test),
+                         numpy.count_nonzero(self.ratings_matrix))
+        self.assertTrue(numpy.all((train * test) == 0))
```
Comment (ndomar) [inline tests/collaborative_tests.py]: No need for this test case if you return the line in the split method.

File: lib/evaluator.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 and evaluates the given recommender.
```
Comment (ndomar) [inline lib/evaluator.py]: In general, ndcg, mrr, and recall_at_x methods should be combined into one. The computationally expensive part is calculating top recommendations and they are always the same for every user so it doesn't make sense to do it 3 times. If you opt to write a helper method that calculates top recommendations for every user and passes it to the 3 methods, you will add o(kn) space complexity. So can we merge them in one method please?

File: lib/evaluator.py [hunk 4]
```diff
@@ -50,25 +51,79 @@ def calculate_recall(self, ratings, predictions):
...
+            top_recommendations = TopRecommendations(n_recommendations)
+            ctr = 0
+            for rating in predictions[user, :]:
+                top_recommendations.insert(ctr, rating)
+                ctr += 1
+            dcg = 0
+            idcg = 0
+            for pos_index, index in enumerate(list(reversed(top_recommendations.get_indices()))):
+                dcg += numpy.power(2, self.ratings[user, index]) - 1 / numpy.log2(pos_index + 2)
+            for pos_index, rating in enumerate(sorted(self.ratings[user, :], reverse=True)):
```
Comment (ndomar) [inline lib/evaluator.py]: The whole idea of calculating top recommendations is to avoid sorting. Here, you can use numpy nonzeros method https://docs.scipy.org/doc/numpy/reference/generated/numpy.nonzero.html

File: lib/evaluator.py [hunk 5]
```diff
@@ -50,25 +51,79 @@ def calculate_recall(self, ratings, predictions):
...
+        ndcgs = []
+        for user in range(self.ratings.shape[0]):
+            top_recommendations = TopRecommendations(n_recommendations)
+            ctr = 0
+            for rating in predictions[user, :]:
+                top_recommendations.insert(ctr, rating)
+                ctr += 1
+            dcg = 0
+            idcg = 0
+            for pos_index, index in enumerate(list(reversed(top_recommendations.get_indices()))):
```
Comment (ndomar) [inline lib/evaluator.py]: I would prefer a better way of looping instead of reversing the list. for i in range(len(a)-1, -1,-1) should do the job.

File: Makefile [hunk 6]
```diff
@@ -56,4 +56,4 @@ coverage: ## check code coverage quickly with the default Python
 	python-coverage report -m > coverage_report.txt
 	cat coverage_report.txt
 	python-coverage html
-	open htmlcov/index.html
+	open htmlcov/index.html
```
Comment (ibralibya) [inline Makefile]: without it i get "Makefile:59: *** missing separator.  Stop."

File: tests/evaluator_tests.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+        def mock_get_ratings_matrix(self=None):
+
+            return [[int(not bool((article + user) % 3)) for article in range(documents_cnt)]
+                    for user in range(users_cnt)]
+        self.ratings_matrix = numpy.array(mock_get_ratings_matrix())
+        setattr(DataParser, "get_ratings_matrix", mock_get_ratings_matrix)
+
+        evaluator = Evaluator(self.ratings_matrix)
+        config = {'n_factors': 5, '_lambda': 0.01}
```
Comment (ibralibya) [inline tests/evaluator_tests.py]: evaluator is used to initialize collaborative filtering

Comment (ndomar) [inline tests/evaluator_tests.py]: I meant the config variable, it is never used. I ran the tests without them and they work fine. you have two config variables, one is instance variable and the other is this one. this one is never used.

File: lib/evaluator.py [hunk 8]
```diff
@@ -4,6 +4,7 @@
 and evaluates the given recommender.
 """
 import numpy
+import math
```
Comment (ndomar) [inline lib/evaluator.py]: Can we remove this? it is not used.

File: tests/collaborative_tests.py [hunk 9]
```diff
@@ -51,6 +51,10 @@ def runTest(self):
         random_item = int(numpy.random.random() * self.documents)
         random_prediction = cf.predict(random_user, random_item)
         self.assertTrue(isinstance(random_prediction, numpy.float64))
+        train, test = cf.naive_split(test_percentage=0.2)
```
Comment (ndomar) [inline tests/collaborative_tests.py]: test_percentage is by default 0.2 so I think this is not needed

File: tests/evaluator_tests.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+                    self.assertEqual(self.rounded_predictions[row, col], self.ratings_matrix[row, col])
+
+        # If we modify all the top predictions for half the users,
+        # recall should be 0.5 by definition
+        for i in range(0, self.users, 2):
+            evaluator.ratings[i, (numpy.argmax(self.predictions[i], axis=0))] = 0
+        recall_at_x = evaluator.recall_at_x(self.n_recommendations, self.predictions)
+        self.assertEqual(0.5, recall_at_x)
+
+        self.setUp()
```
Comment (ndomar) [inline tests/evaluator_tests.py]: The setup function is automatically called before any other function is ran, why is this necessary?
